ksql-engine/src/main/java/io/confluent/ksql/util/ExecutorUtil.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.util;
+
+import org.apache.kafka.common.errors.RetriableException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.function.Supplier;
+
+public final class ExecutorUtil {
+
+  private static final int NUM_RETRIES = 5;
+  private static final int RETRY_BACKOFF_MS = 500;
+  private static final Logger log = LoggerFactory.getLogger(ExecutorUtil.class);
+
+  private ExecutorUtil() {
+  }
+
+  public enum RetryBehaviour {
+    ALWAYS,
+    ON_RETRYABLE
+  }
+
+  @FunctionalInterface
+  public interface Function {
+    void call() throws Exception;
+  }
+
+  public static void execute(final Function function,
+                             final RetryBehaviour retryBehaviour) throws Exception {
+    execute(() -> {
+      function.call();
+      return null;
+    }, retryBehaviour);
+  }
+
+  public static <T> T execute(final Callable<T> supplier,
+                              final RetryBehaviour retryBehaviour) throws Exception {
+    return executeWithRetries(() -> {
+      final CompletableFuture<T> f = new CompletableFuture<>();
+      try {
+        final T result = supplier.call();
+        f.complete(result);
+      } catch (final Exception e) {
+        f.completeExceptionally(e);
+      }
+      return f;
+    }, retryBehaviour);
+  }
+
+  public static <T> T executeWithRetries(final Supplier<? extends Future<T>> supplier,
```

Comment:
Hey @uurl, you've missed the bit from @rodesai where he correctly points out that this executor code doesn't need to worry about `Future`s.

if you change this code to something like:

```
public static void executeWithRetries(final Function function,
                                        final RetryBehaviour retryBehaviour) throws Exception {
    executeWithRetries(() -> {
      function.call();
      return null;
    }, retryBehaviour);
  }

  public static <T> T executeWithRetries(final Callable<T> supplier,
                                         final RetryBehaviour retryBehaviour) throws Exception {
    Exception lastException = null;
    for (int retries = 0; retries < NUM_RETRIES; ++retries) {
      try {
        if (retries != 0) {
          Thread.sleep(RETRY_BACKOFF_MS);
        }
        return supplier.call();
      } catch (final Exception e) {
        final Throwable cause = e instanceof ExecutionException ? e.getCause() : e;

        if (cause instanceof RetriableException
            || (cause instanceof Exception && retryBehaviour == RetryBehaviour.ALWAYS)) {
          log.info("Retrying request. Retry no: " + retries, e);
          lastException = e;
        } else if (cause instanceof Exception) {
          throw (Exception) cause;
        } else {
          throw new RuntimeException(e.getMessage());
        }
      }
    }
    throw lastException;
  }
```

Then you can change the calling code that uses futures to, for example:

```
ExecutorUtil.executeWithRetries(
         () -> adminClient.createTopics(Collections.singleton(newTopic)).all().get(),
          ExecutorUtil.RetryBehaviour.ON_RETRYABLE);
```

Note the additional `.get()` on the end of the lambda.  If this get throws an exception, then the lambda throws and the `ExecutorUtil` class will catch it and handle it. So there is no need for the Util class to have `Future` overloads.  Nice hey!

Notice I've renamed all methods to `executeWithReties`.

With this change you can update the tests by removing all the references to futures.